File: concurrent-programming-with-monitors/QueueTwoConds.java
/* 
 * Concurrent programming with monitors (Java)
 * by Project Nayuki, 2024. Public domain.
 * https://www.nayuki.io/page/concurrent-programming-with-monitors
 */

import java.util.LinkedList;
import java.util.Random;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;


public final class QueueTwoConds {
	
	// Customizable parameters
	private static final int NUM_THREADS = 4;
	private static final int MAX_DELAY_MS = 10_000;
	
	
	// State variables
	private static final long startTime = System.currentTimeMillis();
	private static Queue<Integer> queue = new Queue<>(3);
	
	
	public static void main(String[] args) {
		for (int i = 0; i < NUM_THREADS; i++)
			new Thread(QueueTwoConds::worker, "worker" + i).start();
	}
	
	
	private static void timestampedPrintf(String format, Object... args) {
		long time = System.currentTimeMillis() - startTime;
		var newArgs = new Object[args.length + 2];
		newArgs[0] = time;
		newArgs[1] = Thread.currentThread().getName();
		System.arraycopy(args, 0, newArgs, 2, args.length);
		synchronized(QueueTwoConds.class) {
			System.out.printf("[%5d ms] %s | " + format, newArgs);
		}
	}
	
	
	private static void worker() {
		try {
			while (true) {
				int duration = rand.nextInt(MAX_DELAY_MS);
				timestampedPrintf("Sleep %d ms%n", duration);
				Thread.sleep(duration);
				
				if (rand.nextDouble() < 0.5) {
					int val = rand.nextInt(10000);
					queue.add(val);
				} else {
					queue.remove();
				}
			}
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
	
	
	private static Random rand = new Random();
	
	
	
	private static final class Queue<E> {
		
		private final int limit;
		private Lock lock = new ReentrantLock();
		private Condition canAdd = lock.newCondition();
		private Condition canRemove = lock.newCondition();
		private LinkedList<E> data = new LinkedList<>();
		
		
		public Queue(int limit) {
			if (limit < 1)
				throw new IllegalArgumentException();
			this.limit = limit;
		}
		
		
		public void add(E val) throws InterruptedException {
			timestampedPrintf("val=%d -> add() -> begin%n", val);
			lock.lock();
			try {
				while (data.size() == limit) {
					timestampedPrintf("add() -> length=%d -> wait%n", data.size());
					canAdd.await();
				}
				data.add(val);
				canRemove.signal();
				timestampedPrintf("add() -> length=%d -> length=%d -> end%n", data.size() - 1, data.size());
			} finally {
				lock.unlock();
			}
		}
		
		
		public E remove() throws InterruptedException {
			timestampedPrintf("remove() -> begin%n");
			lock.lock();
			try {
				while (data.isEmpty()) {
					timestampedPrintf("remove() -> length=%d -> wait%n", data.size());
					canRemove.await();
				}
				E result = data.remove();
				canAdd.signal();
				timestampedPrintf("remove() -> val=%d -> length=%d -> length=%d -> end%n", result, data.size() + 1, data.size());
				return result;
			} finally {
				lock.unlock();
			}
		}
		
	}
	
}

File: concurrent-programming-with-monitors/queue-two-conds.py
# 
# Concurrent programming with monitors (Python)
# by Project Nayuki, 2024. Public domain.
# https://www.nayuki.io/page/concurrent-programming-with-monitors
# 

from __future__ import annotations
import collections, random, threading, time
from typing import Generic, TypeVar


# Customizable parameters
_NUM_THREADS: int = 4
_MAX_DELAY_S: float = 10.0


# State variables
_start_time: float = time.time()
_queue: _Queue[int]


def main() -> None:
	global _queue
	_queue = _Queue(3)
	for i in range(_NUM_THREADS):
		threading.Thread(name=f"Worker{i}", target=_worker).start()


def _worker() -> None:
	while True:
		duration: float = random.uniform(0.0, _MAX_DELAY_S)
		_timestamped_print(f"Sleep {round(duration * 1000)} ms")
		time.sleep(duration)
		
		if random.random() < 0.5:
			val: int = random.randrange(10000)
			_queue.add(val)
		else:
			_queue.remove()


_print_lock: threading.Lock = threading.Lock()

def _timestamped_print(s: str) -> None:
	relativetime: float = time.time() - _start_time
	s = f"[{int(relativetime * 1000):5d} ms] {threading.current_thread().name} | {s}"
	with _print_lock:
		print(s)



_E = TypeVar("_E")

class _Queue(Generic[_E]):
	
	_limit: int
	_lock: threading.Lock
	_can_add: threading.Condition
	_can_remove: threading.Condition
	_data: collections.deque[_E]
	
	
	def __init__(self, limit: int):
		if limit < 1:
			raise ValueError()
		self._limit = limit
		self._lock = threading.Lock()
		self._can_add = threading.Condition(self._lock)
		self._can_remove = threading.Condition(self._lock)
		self._data = collections.deque()
	
	
	def add(self, val: _E) -> None:
		_timestamped_print(f"val={val} -> add() -> begin")
		with self._lock:
			while len(self._data) == self._limit:
				_timestamped_print(f"add() -> length={len(self._data)} -> wait")
				self._can_add.wait()
			self._data.append(val)
			self._can_remove.notify()
			_timestamped_print(f"add() -> length={len(self._data)-1} -> length={len(self._data)} -> end")
	
	
	def remove(self) -> _E:
		_timestamped_print("remove() -> begin")
		with self._lock:
			while len(self._data) == 0:
				_timestamped_print(f"remove() -> length={len(self._data)} -> wait")
				self._can_remove.wait()
			result: _E = self._data.popleft()
			self._can_add.notify()
			_timestamped_print(f"remove() -> val={result} -> length={len(self._data)+1} -> length={len(self._data)} -> end")
			return result



if __name__ == "__main__":
	main()
